### PR TITLE
(PA-2941) Setup testing environment for beaker-gke

### PR DESCRIPTION
**Pull request by luchihoratiu**

### CODE REVIEW

File: beaker-gke.gemspec [hunk 1]
```diff
@@ -19,9 +19,16 @@ Gem::Specification.new do |s|
...
 
-  # required ruby version
+  # Required ruby version
   s.required_ruby_version = '~> 2.3'
 
+  # Testing dependencies
+  s.add_development_dependency 'fakefs', '~> 0.6', '<= 0.13.3'
+  s.add_development_dependency 'rake', '~> 10.1'
+  s.add_development_dependency 'rspec', '~> 3.0'
+  s.add_development_dependency 'rspec-its'
```
Comment (gimmyxd) [inline beaker-gke.gemspec:30]: I think we can drop this, we're not gonna write tests with `its`

File: spec/gke/hypervisor/gke_spec.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'fakefs/spec_helpers'
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: you can create a `.rspec` file in the root directory and add 
```
require 'spec_helper'
```
`require 'fakefs/spec_helpers'`  could be required in `spec_helper.rb` if we're gonna use it in each test file.

and any other rspec config. In this way we won't have to require those in each test file.
https://github.com/rspec/rspec/wiki#rspec

File: spec/gke/hypervisor/gke_spec.rb [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  let(:hosts) do
+    the_hosts = make_hosts
+    the_hosts
+  end
+
+  let(:options) do
+    {
+      logger: logger
+    }
+  end
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: nit: you can compact this on single line
```ruby
let(options) { { logger: logger } }
```

File: spec/gke/hypervisor/gke_spec.rb [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  before do
+    FakeFS.deactivate!
+  end
+
+  describe ' # initialize' do
+    let(:env_error_message) { 'OS environment variable KUBECONFIG and GOOGLE_APPLICATION_CREDENTIALS must be set' }
+
+    it 'raises error when KUBECONFIG and GOOGLE_APPLICATION_CREDENTIALS ENV variables are not set' do
+      allow(ENV).to receive(:fetch).with('GOOGLE_APPLICATION_CREDENTIALS').and_raise(KeyError)
+      allow(ENV).to receive(:fetch).with('KUBECONFIG').and_raise(KeyError)
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: This are not needed as `Hash.fetch` raises KeyError if key is not found.
https://apidock.com/ruby/Hash/fetch

File: spec/gke/hypervisor/gke_spec.rb [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+      expect { gke }.to raise_error(ArgumentError, env_error_message)
+    end
+
+    it 'raises error when KUBECONFIG ENV variable is not set' do
+      allow(ENV).to receive(:fetch).with('GOOGLE_APPLICATION_CREDENTIALS').and_return('default_value')
+      allow(ENV).to receive(:fetch).with('KUBECONFIG').and_raise(KeyError)
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: not needed as per above comment

File: spec/gke/hypervisor/gke_spec.rb [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+      expect { gke }.to raise_error(ArgumentError, env_error_message)
+    end
+
+    it 'raises error when GOOGLE_APPLICATION_CREDENTIALS ENV variable is not set' do
+      allow(ENV).to receive(:fetch).with('KUBECONFIG').and_return('default_value')
+      allow(ENV).to receive(:fetch).with('GOOGLE_APPLICATION_CREDENTIALS').and_raise(KeyError)
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: not needed as per above comment

File: spec/gke/hypervisor/gke_spec.rb [hunk 7]
```diff
@@ -0,0 +1,178 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'fakefs/spec_helpers'
+
+describe Beaker::Gke do
+  let(:hosts) do
+    the_hosts = make_hosts
+    the_hosts
+  end
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: You can just call make_hosts here

File: spec/gke/hypervisor/gke_spec.rb [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      expect { gke }.to raise_error(ArgumentError, env_error_message)
+    end
+
+    it 'does not raise ArgumentError when KUBECONFIG and GOOGLE_APPLICATION_CREDENTIALS ENV are set' do
+      RSpec::Expectations.configuration.on_potential_false_positives = :nothing
+      allow(ENV).to receive(:fetch).with('GOOGLE_APPLICATION_CREDENTIALS').and_return('default_value')
+      allow(ENV).to receive(:fetch).with('KUBECONFIG').and_return('default_value')
+
+      expect { gke }.not_to raise_error(ArgumentError, env_error_message)
+    end
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: This is passing, even though @client  = client(https://github.com/puppetlabs/beaker-gke/blob/master/lib/beaker/hypervisor/gke.rb#L30) will try to create an instance of kubeclient.

I suggest to drop  `expect { gke }.not_to raise_error(ArgumentError, env_error_message)` and test that the instance variables are set correctly, eg: `expect(gke.hosts).to eql(hosts)`

File: spec/gke/hypervisor/gke_spec.rb [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    allow(ENV).to receive(:fetch).with('GOOGLE_APPLICATION_CREDENTIALS').and_return('default_value')
+    allow(config).to receive(:context).and_return(context)
+    allow(Kubeclient::Config).to receive(:read).with(ENV['KUBECONFIG']).and_return(config)
+  end
+
+  before do
+    FakeFS.deactivate!
+  end
+
+  describe ' # initialize' do
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb]: nit: usually this is written like: ```describe ' #initialize' do```

File: spec/gke/hypervisor/gke_spec.rb [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    context 'when no hosts given' do
+      let(:no_hosts_gke) { ::Beaker::Gke.new([], options) }
+
+      it 'does not raise any error' do
+        expect { no_hosts_gke.cleanup }.not_to raise_error
+      end
+
+      it 'logs no info' do
+        expect(logger).not_to have_received(:info)
```
Comment (gimmyxd) [inline spec/gke/hypervisor/gke_spec.rb:154]: I think you must call `no_hosts_gke.cleanup` before this expect

File: .rubocop.yml [hunk 11]
```diff
@@ -20,3 +20,16 @@ Metrics/AbcSize:
...
+    - 'Gemfile'
+
+Metrics/BlockLength:
+  Enabled: false
+
+Metrics/ModuleLength:
+  Enabled: false
+
+RSpec/FilePath:
+  Enabled: false
```
Comment (gimmyxd) [inline .rubocop.yml]: nit: no newline at end of file

File: beaker-gke.gemspec (hunk 1 above)
Comment (luchihoratiu) [reply beaker-gke.gemspec:30]: I don't think we can. This seems to be required by spec_helper of Beaker.

> An error occurred while loading spec_helper. - Did you mean?
    rspec ./spec/spec_helper.rb
Failure/Error: require File.join(beaker_spec_path, 'spec_helper.rb')
LoadError:
  cannot load such file -- rspec/its
.rvm/gems/ruby-2.6.3/gems/beaker-4.13.1/spec/spec_helper.rb:10:in `require'
.rvm/gems/ruby-2.6.3/gems/beaker-4.13.1/spec/spec_helper.rb:10:in `<top (required)>'
./spec/spec_helper.rb:14:in `require'
./spec/spec_helper.rb:14:in `<top (required)>'
No examples found.

**Review by gimmyxd — APPROVED** [review]

**Pull request merged**